sdk/core/Azure.Core/src/ResourceIdentifier.cs

```diff
@@ -0,0 +1,429 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using System.Text;
+
+namespace Azure.Core
+{
+    /// <summary>
+    /// An Azure Resource Manager resource identifier.
+    /// </summary>
+    public sealed class ResourceIdentifier : IEquatable<ResourceIdentifier>, IComparable<ResourceIdentifier>
+    {
+        private const char Separator = '/';
+
+        private static ResourceType TenantResourceType = new ResourceType("Microsoft.Resources/tenants");
+        private static ResourceType SubscriptionResourceType = new ResourceType("Microsoft.Resources/subscriptions");
+        private static ResourceType ResourceGroupResourceType = new ResourceType("Microsoft.Resources/resourceGroups");
+        private static ResourceType ProviderResourceType = new ResourceType("Microsoft.Resources/providers");
+        private const string RootStringValue = "/";
+        private const string ProvidersKey = "providers";
+        private const string SubscriptionsKey = "subscriptions";
+        private const string LocationsKey = "locations";
+        private const string ResourceGroupsLowerKey = "resourcegroups";
+
+        private string? _stringValue;
+
+        /// <summary>
+        /// The root of the resource hierarchy.
+        /// </summary>
+        public static readonly ResourceIdentifier Root = new ResourceIdentifier(null, TenantResourceType, string.Empty, true);
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ResourceIdentifier"/> class.
+        /// </summary>
+        /// <param name="resourceId"> The id string to create the ResourceIdentifier from. </param>
+        public ResourceIdentifier(string resourceId)
+        {
+            var id = Create(resourceId);
```

Comment:
Nit: use explicit type instead of `var` here for clarity.

Reply:
It looks like we are creating an instance of a resource identifier and then copying it into the one we're constructing in this constructor.  Could we find a way to do this without creating two instances of ResourceIdentifier?